sdk/test/ProtocolTests/Generated/JSONRPC10/GreetingWithErrors.cs

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ * 
+ *  http://aws.amazon.com/apache2.0
+ * 
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * Do not modify this file. This file is generated.
+ */
+using AWSSDK.ProtocolTests;
+using AWSSDK.ProtocolTests.Utils;
+using AWSSDK_DotNet35.UnitTests.TestTools;
+using Amazon.JSONRPC10;
+using Amazon.JSONRPC10.Model;
+using Amazon.JSONRPC10.Model.Internal.MarshallTransformations;
+using Amazon.Runtime;
+using Amazon.Runtime.Internal;
+using Amazon.Runtime.Internal.Transform;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text;
+
+namespace AWSSDK.ProtocolTests.JsonRpc10
+{
+    [TestClass]
+    public class GreetingWithErrors
+    {
+        /// <summary>
+        /// Parses simple JSON errors
+        /// </summary>
+        [TestMethod]
+        [TestCategory("ProtocolTest")]
+        [TestCategory("ErrorTest")]
+        [TestCategory("JsonRpc10")]
+        public void AwsJson10InvalidGreetingErrorErrorResponse()
+        {
+            // Arrange
+            var webResponseData = new WebResponseData();
+            webResponseData.StatusCode = (HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), 400);
+            webResponseData.Headers["Content-Type"] = "application/x-amz-json-1.0";
+            byte[] bytes = Encoding.ASCII.GetBytes("{\n    \"__type\": \"aws.protocoltests.json10#InvalidGreeting\",\n    \"Message\": \"Hi\"\n}");
+            var stream = new MemoryStream(bytes);
+            var context = new JsonUnmarshallerContext(stream,true,webResponseData);
+            // Act
+            var errorResponse = new GreetingWithErrorsResponseUnmarshaller().UnmarshallException(context, new InvalidGreetingException(""), (HttpStatusCode)Enum.ToObject(typeof(HttpStatusCode), 400));
```

Comment:
I'm not sure how other SDK's handle errors but we just take the status code that the service returns and either return the specific exception (if it can be determined by the headers and other data in the payload) along with the message, status code, and inner exceptio etc. or if that can't be determined we return the service-level exception. The expected exception is there to return as the inner exception of the parent exception.

For the purposes of the test, I had to pass in this data because the unmarshallException method expects them.

Reply:
offline we talked about:
* 2nd parameter being `innerException`. Can avoid using the same exception type as the expected exception test of this test (i.e., InvalidGreetingException) and instead use a generic runtime exception or null if that's allowed.
* 3rd parameter: i'd imagine the unmarshaller is eventually setting errorResponse.statusCode based on WebResponseData in the 1st method parameter OR this explicit value in 3rd parameter. It seems that currently the 3rd parameter is getting used and not WebResponseData. You could see if you'd like to change that in your Unmarshaller logic itself. But maybe fine to set the value in both WebResponseData and this 3rd parameter, so you can change the Unmarshaller later without affecting this test.
  * But I'd say either ways, add an assert for errorResponse.statusCode == 400

Reply:
without more context, it seems here https://github.com/aws/aws-sdk-net/blob/23b74540be451eda8bb789980b4863bd8cc4241f/sdk/src/Core/Amazon.Runtime/Internal/Transform/JsonErrorResponseUnmarshaller.cs#L133 this might have worked `response.StatusCode = context.ResponseData.StatusCode` and avoided the 3rd parameter to the method

Reply:
I decided to go with the 2nd parameter being null. It seems like the 3rd parameter is being used and not WebResponseData, but I'd like to change the unmarshaller after the fact and not change the test. The test works as expected with the 2nd parameter being null though